src/realm/object-store/c_api/socket_provider.cpp

```diff
@@ -0,0 +1,250 @@
+
+    ~CAPISyncSocketProvider()
+    {
+        m_free(m_userdata);
+    }
+
+    std::unique_ptr<sync::WebSocketInterface> connect(sync::WebSocketObserver* observer,
+                                                      sync::WebSocketEndpoint&& endpoint) final
+    {
+        auto capi_observer = std::make_shared<CAPIWebSocketObserver>(*observer);
+        return std::make_unique<CAPIWebSocket>(m_userdata, m_websocket_connect, m_websocket_async_write,
+                                               m_websocket_free, new realm_websocket_observer_t(capi_observer),
+                                               std::move(endpoint));
+    }
+
+    void post(FunctionHandler&& handler) final
+    {
+        auto shared_handler = std::make_shared<FunctionHandler>(std::move(handler));
+        m_post(m_userdata, new realm_sync_socket_callback_t(std::move(shared_handler)));
+    }
+
+    SyncTimer create_timer(std::chrono::milliseconds delay, FunctionHandler&& handler) final
+    {
+        auto shared_handler = std::make_shared<FunctionHandler>(std::move(handler));
+        return std::make_unique<CAPITimer>(m_userdata, delay.count(),
+                                           new realm_sync_socket_callback_t(std::move(shared_handler)),
+                                           m_timer_create, m_timer_cancel, m_timer_free);
+    }
+};
+
+} // namespace
+
+RLM_API realm_sync_socket_t* realm_sync_socket_new(
+    realm_userdata_t userdata, realm_free_userdata_func_t userdata_free, realm_sync_socket_post_func_t post_func,
+    realm_sync_socket_create_timer_func_t create_timer_func,
+    realm_sync_socket_timer_canceled_func_t cancel_timer_func, realm_sync_socket_timer_free_func_t free_timer_func,
+    realm_sync_socket_connect_func_t websocket_connect_func,
+    realm_sync_socket_websocket_async_write_func_t websocket_write_func,
+    realm_sync_socket_websocket_free_func_t websocket_free_func)
+{
+    return wrap_err([&]() {
+        auto capi_socket_provider = std::make_shared<CAPISyncSocketProvider>();
+        capi_socket_provider->m_userdata = userdata;
+        capi_socket_provider->m_free = userdata_free;
+        capi_socket_provider->m_post = post_func;
+        capi_socket_provider->m_timer_create = create_timer_func;
+        capi_socket_provider->m_timer_cancel = cancel_timer_func;
+        capi_socket_provider->m_timer_free = free_timer_func;
+        capi_socket_provider->m_websocket_connect = websocket_connect_func;
+        capi_socket_provider->m_websocket_async_write = websocket_write_func;
+        capi_socket_provider->m_websocket_free = websocket_free_func;
+        return new realm_sync_socket_t(std::move(capi_socket_provider));
+    });
+}
+
+RLM_API void realm_sync_socket_callback_complete(realm_sync_socket_callback* realm_callback,
+                                                 status_error_code_e status, const char* reason)
+{
+    (*(realm_callback->get()))(Status{static_cast<ErrorCodes::Error>(status), reason});
```

Comment:
For `status_error_code_e` of `OK`, we'll need to pass `Status::OK()` instead of creating a `Status` object.

Reply:
good catch